### PR TITLE
Add gradcheck diagnostic for dependent checked inputs

### PR DESCRIPTION
**Pull request by GeorgePatel**
# Add gradcheck diagnostic for dependent checked inputs

Fixes #121842.

## Summary

This PR documents and diagnoses a class of `gradcheck` failures that can occur when one checked input is computed from another checked input.

For example:

```python
x = torch.ones(1, dtype=torch.double, requires_grad=True)
y = 3 * x

gradcheck(torch.dot, (x, y))
```

currently reports:

```text
numerical:  3
analytical: 6
```

The backward implementation is not necessarily incorrect in this situation. The two sides of `gradcheck` can be differentiating against different effective input boundaries:

* **Numerical differentiation** perturbs one explicit checked input at a time without recomputing the other checked inputs from it.
* **Backward-mode autograd** follows the pre-existing autograd graph connecting those tensors.

> **Note**  
> This PR does **not** automatically rewrite or detach user inputs.

Instead, it:
1. Documents the distinction in the `gradcheck` API documentation.
2. Adds a targeted diagnostic after a backward-mode numerical/analytical Jacobian mismatch when the mismatched input has a potential graph dependency to another checked input.

## Why the mismatch occurs

Suppose the function being checked is:
$$F(x, y)$$

but the concrete tensors passed to `gradcheck` satisfy:
$$y = g(x)$$

Numerically, `gradcheck` perturbs $x$ while leaving the explicit $y$ argument unchanged. Ignoring storage aliasing, this treats the supplied arguments as independent coordinates.

For the original reproducer,
$$F(x,y)=xy, \qquad y=3x$$

the numerical derivative with respect to $x$ is therefore:
$$\frac{\partial F}{\partial x}=y=3$$

Backward autograd sees the existing graph:

```mermaid
graph LR
    x --> F
    x --> y
    y --> F
```

so the gradient with respect to $x$ can include both paths:

$$ \frac{dF}{dx} = \frac{\partial F}{\partial x} + \frac{\partial F}{\partial y}\frac{dy}{dx} $$

At $x = 1$, $y = 3$:

$$3 + 1 \cdot 3 = 6$$

This is the `3` versus `6` mismatch reported in #121842.

The same issue is transitive. For example, an input relationship such as:

```mermaid
graph LR
    x --> sin --> intermediate --> multiply --> z
```

can produce the same mismatch when both $x$ and $z$ are supplied as checked inputs.

By contrast, a shared ancestor alone is not sufficient:

```mermaid
graph TD
    h --> q
    h --> k
```

Neither $q$ nor $k$ is upstream of the other, so checking a function of $(q, k)$ does not introduce the same additional path into either requested derivative.

## Why this PR does not automatically detach inputs

A possible implementation-level solution is to detach checked inputs before computing the analytical Jacobian.

That would make the explicit arguments independent autograd coordinates, but it would require `gradcheck` to silently choose a mathematical interpretation that it cannot infer from the call.

Given:

```python
gradcheck(F, (x, y))
```

with $y = g(x)$, the caller may intend either of two different tests.

### Local Jacobian

The caller may want to test:
$$F(x,y)$$

as a function of independent coordinates.

In that case, the appropriate setup is to sever the pre-existing dependency before calling `gradcheck`, for example with detached inputs when appropriate.

### Composed function

The caller may instead want to test:
$$x \mapsto F(x,g(x))$$

In that case, the dependency is part of the function being checked and should be reconstructed inside the callable:

```python
gradcheck(lambda x: F(x, g(x)), (x,))
```

Now numerical perturbations of `x` also recompute `g(x)`, so numerical and analytical differentiation operate on the same composition.

`gradcheck(F, (x, y))` does not contain enough information to distinguish these intentions. 

Automatic detachment would therefore silently select the local-Jacobian interpretation. It also changes autograd object semantics: a differentiable non-leaf tensor becomes a leaf after `detach().requires_grad_(True)`. That can affect otherwise valid leaf-sensitive behavior such as custom `autograd.Function`s that legitimately perform in-place operations using `mark_dirty`.

For these reasons, this PR does not mutate checked inputs.

## Why the diagnostic runs only after a mismatch

A graph relationship alone does not prove that it caused a Jacobian mismatch.

For checked inputs $I_i$ and $I_j$, the additional analytical contribution has the form:
$$\frac{\partial F}{\partial I_j} \frac{dI_j}{dI_i}$$

Graph reachability can indicate that the second factor may exist, but it does not prove that the complete term is nonzero.

For example:

```mermaid
graph LR
    x --> y
```

may exist while the tested function does not depend on `y`. In that case, the dependency is harmless for the current derivative.

A preflight rejection or automatic intervention based only on graph topology would therefore affect cases that currently pass.

Instead, the dependency traversal runs only after `gradcheck` has already observed a backward-mode numerical/analytical Jacobian mismatch. The diagnostic is also scoped to the specific input whose Jacobian mismatched, so a dependency between unrelated checked inputs does not decorate an unrelated failure.

Successful `gradcheck` calls remain on their existing execution paths.

## Why the diagnostic says "potential dependency"

The graph traversal deliberately overapproximates mathematical dependence.

`next_functions` belongs to an autograd `Node`, not to an individual output of that node. 

For example, consider a custom multi-output `Function`:

```python
class SplitDependency(torch.autograd.Function):
    @staticmethod
    def forward(ctx, a, b):
        return a.clone(), b.clone()

    @staticmethod
    def backward(ctx, grad_a, grad_b):
        return grad_a, grad_b
```

Mathematically, the second output does not depend on the first input:
$$\frac{\partial\,out_b}{\partial a}=0$$

However, both outputs share the same backward `Node`, whose `next_functions` exposes the upstream edges for the function invocation as a whole. 

A lightweight graph traversal can therefore observe structural reachability from `a` through that node even though the relevant output-specific derivative is zero. This cannot be resolved by replacing BFS with DFS or a topological sort: the missing information is output-specific derivative dependence, not graph-search order.

The diagnostic therefore intentionally says that the graph contains a **potential dependency** and that such a relationship **can** explain the mismatch. It does not claim that reachability proves causation.

## User-facing behavior

For the original #121842 reproducer, the existing error:

```text
Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[3.0000]], dtype=torch.float64)
analytical:tensor([[6.]], dtype=torch.float64)
```

is followed by:

```text
The autograd graph contains a potential dependency from this input to another
checked input. Numerical gradcheck perturbs one checked input at a time without
recomputing the other explicit inputs from it, whereas backward mode autograd can
additionally propagate through such a dependency and pick up extra chain rule
terms. That can make the numerical and analytical checks disagree even when the
backward implementation is correct. To check the composed function, build the
dependency inside func, for example gradcheck(lambda x: func(x, g(x)), (x,)). To
check local Jacobians, pass inputs that do not depend on each other, for example
x.detach().requires_grad_().
```

The `gradcheck` documentation contains the corresponding longer explanation of the two possible testing intents.

## Alternatives considered

**Automatically detach checked inputs**  
Rejected as the default behavior because it silently chooses local-Jacobian semantics, changes non-leaf tensors into leaves, and can alter valid autograd behavior unrelated to this issue.

**Reject dependent inputs before running gradcheck**  
Rejected because graph reachability is only a structural risk condition. The relevant chain-rule contribution can be zero, so some such inputs currently pass correctly.

**Documentation only**  
This would address discoverability but would leave users encountering the original failure without context at the point where the ambiguity becomes relevant.

This PR therefore uses documentation plus a post-mismatch diagnostic: the execution semantics remain unchanged, while the failure explains the two possible differentiation boundaries.

## Testing

New coverage includes:
- original #121842 direct dependency;
- transitive dependency;
- shared-ancestor inputs that must continue to pass;
- genuine backward mismatches with no dependency hint;
- dependency relationships unrelated to the mismatched input;
- `(node, output_nr)` gradient-edge identity;
- duplicate checked inputs;
- nested checked inputs;
- documented multi-output overapproximation;
- forward-only AD;
- `gradgradcheck`;
- `raise_exception=False`;
- both `fast_mode=True` and `fast_mode=False`.

Local validation:

```bash
python test/test_autograd.py TestAutograd -k gradcheck -v
```
Result: 32 passed, 4 skipped.

```bash
python test/test_autograd.py
```
Result: 770 passed, 50 skipped, 1 error.

```bash
lintrunner -a
git diff --check
```
Both are clean.

## Backward compatibility

This PR does not change the numerical or analytical computation performed by successful gradchecks and does not mutate user inputs.

The only runtime behavior change is additional explanatory text on an existing backward-mode Jacobian mismatch when the mismatched input has a potential dependency to another checked input.

Because exception text is observable behavior, this is technically a user-visible diagnostic change, but it does not change whether an existing check succeeds or fails.

### CODE REVIEW

**Comment by GeorgePatel** [comment]
@soulitzer @albanD gentle ping for review when you have a chance. This PR addresses actionable #121842 by documenting the dependent-input behavior in gradcheck and adding a targeted diagnostic without changing successful gradcheck semantics. Happy to address any feedback. Thanks!
